src/System.Management.Automation/engine/Subsystem/Prediction/CommandPrediction.cs

```diff
@@ -0,0 +1,174 @@
+    /// Provides a set of possible predictions for given input.
+    /// </summary>
+    public static class CommandPrediction
+    {
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the default timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens)
+        {
+            return PredictInput(ast, astTokens, millisecondsTimeout: 20);
+        }
+
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the specified timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <param name="millisecondsTimeout">The milliseconds to timeout.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static async Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens, int millisecondsTimeout)
+        {
+            var predictors = SubsystemManager.GetSubsystems<IPredictor>();
+            if (predictors.Count == 0)
+            {
+                return null;
+            }
+
+            var context = new PredictionContext(ast, astTokens);
+            var cancellationSource = new CancellationTokenSource();
+            var cancellationToken = cancellationSource.Token;
+            var tasks = new Task<PredictionResult>[predictors.Count];
+
+            for (int i = 0; i < predictors.Count; i++)
+            {
+                var predictor = predictors[i];
+
+                tasks[i] = Task.Factory.StartNew(
+                    state => {
+                        var predictor = (IPredictor)state;
+                        List<string> texts = predictor.GetSuggestion(context, cancellationToken);
+                        return texts?.Count > 0 ? new PredictionResult(predictor.Id, texts) : null;
+                    }, predictor);
+            }
+
+            try
+            {
+                await Task.WhenAny(
+                    Task.WhenAll(tasks),
+                    Task.Delay(millisecondsTimeout, cancellationToken)).ConfigureAwait(false);
+                cancellationSource.Cancel();
+
+                var results = new List<PredictionResult>(predictors.Count);
+                foreach (Task<PredictionResult> task in tasks)
+                {
+                    if (task.IsCompletedSuccessfully)
+                    {
+                        var result = task.Result;
```

Comment:
Consider:

```suggestion
                        var result = await task;
```

Reply:
When reaching here, `task.IsCompletedSuccessfully` is true, so `task.Result` should just return the result that is already available, so I think `await task` is not needed.